# 3. Guided Project: Exploring Ebay Car Sales Data

In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen. 

The data dictionary provided with data is as follows:

* dateCrawled - When this ad was first crawled. All field-values are taken from this date.
* name - Name of the car.
* seller - Whether the seller is private or a dealer.
* offerType - The type of listing
* price - The price on the ad to sell the car.
* abtest - Whether the listing is included in an A/B test.
* vehicleType - The vehicle Type.
* yearOfRegistration - The year in which the car was first registered.
* gearbox - The transmission type.
* powerPS - The power of the car in PS. 
* model - The car model nam* e.
* odometer - How many kilo* meters the car has driven.
* monthOfRegistration - The*  month in which the car was first registered.
* fuelType - What type of.fuel the car uses.
* brand - The brand of the car.
* notRepairedDamage - If the car has a damage which is not yet repaired.
* dateCreated - The date on which the eBay listing was created.
* nrOfPictures - The number of pictures in the ad.
* postalCode - The postal code for the location of the vehicle.
* lastSeenOnline - When the crawler saw this ad last online.

The **aim of this project** is to clean the data and analyze the included used car listings.

In [1]:
# Import the pandas and NumPy libraries

import numpy as np
import pandas as pd

# Read the autos.csv CSV file into pandas, and assign it to the variable name autos.

autos = pd.read_csv("autos.csv", encoding="Latin-1")

# Use the DataFrame.info() and DataFrame.head() methods to print information about the autos dataframe, 
# as well as the first few rows.

autos.info()
autos.head(2)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   

  offerType   price   abtest vehicleType  yearOfRegistration    gearbox  \
0   Angebot  $5,000  control         bus                2004    manuell   
1   Angebot  $8,500  control   limousine                1997  automatik   

   powerPS   model   odometer  monthOfRegistration fuelType    brand  \
0      158  andere  150,000km                    3      lpg  peugeot   
1      286     7er  150,000km                    6   benzin      bmw   

  notRepairedDamage          dateCreated  nrOfPictures  postalCode  \
0              nein  2016-03-26 00:00:00             0       79588   
1              nein  2016-04-04 00:00:00             0       71034   

              lastSeen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08

Observations:

* The dataset contains 20 columns
* Five of the twenty columns have less then 50.000 entries
* The Datatype for fifteen columns is "object" and for the remaining five "integers" 

In [2]:
# Use the DataFrame.columns attribute to print an array of the existing column names.

autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [3]:
col = autos.columns

# Adjustments to the column names (e.g. change camel case against snake case

def clean_col(col):
    col = col.replace("yearOfRegistration","registration_year")
    col = col.replace("monthOfRegistration","registration_month")
    col = col.replace("notRepairedDamage","unrepaired_damage")
    col = col.replace("dateCreated","ad_created")
    col = col.replace("dateCrawled","date_crawled")
    col = col.replace("offerType","offer_type")
    col = col.replace("vehicleType","vehicle_type")
    col = col.replace("powerPS","power_ps")
    col = col.replace("fuelType","fuel_type")
    col = col.replace("notRepairedDamage","not_repaired_damage")
    col = col.replace("nrOfPictures","nr_of_pictures")
    col = col.replace("postalCode","postal_code")
    col = col.replace("lastSeen","last_seen")
    return col

new_columns = []
for c in autos.columns:
    clean_c = clean_col(c)
    new_columns.append(clean_c)

autos.columns = new_columns

# Use DataFrame.head() to look at the current state of the autos dataframe.

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Basic Data Exploration    

In [4]:
autos.describe(include = "all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-12 16:06:22  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

First observations:

1) Any columns that have mostly one value that are candidates to be dropped:

* "seller", "offer_type", "abtest", "gearbox", "unrepaired_damage"

2) Any examples of numeric data stored as text that needs to be cleaned:

* "price", "odometer"

In [5]:
# You likely found that the price and odometer columns are numeric values stored as text. For each column:
# 1) Remove any non-numeric characters.

print(autos["price"].unique())
print(autos["odometer"].unique())

autos["price"] = autos["price"].str.replace("$","")
autos["price"] = autos["price"].str.replace(",","")

autos["odometer"] = autos["odometer"].str.replace("km","")
autos["odometer"] = autos["odometer"].str.replace(",","")

['$5,000' '$8,500' '$8,990' ... '$385' '$22,200' '$16,995']
['150,000km' '70,000km' '50,000km' '80,000km' '10,000km' '30,000km'
 '125,000km' '90,000km' '20,000km' '60,000km' '5,000km' '100,000km'
 '40,000km']


In [6]:
# Did the adjustment work?

print(autos["price"].unique())
print(autos["odometer"].unique())

['5000' '8500' '8990' ... '385' '22200' '16995']
['150000' '70000' '50000' '80000' '10000' '30000' '125000' '90000' '20000'
 '60000' '5000' '100000' '40000']


In [7]:
# 2) Convert the column to a numeric dtype.

autos["price"] = autos["price"].astype(int)
autos["odometer"] = autos["odometer"].astype(int)

In [8]:
# 3) Use DataFrame.rename() to rename the column to odometer_km.

autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)

## Analyze the columns "price" and "odometer"

Find the following values:

1) How many unique values?

2) Min/max/median/mean etc.

3) Value counts

In [9]:
# 1) How many unique values?

print(autos["price"].unique().shape)
print(autos["odometer_km"].unique().shape)

(2357,)
(13,)


In [10]:
# 2) Min/max/median/mean etc.

print(autos["price"].describe())
print("\n")
print(autos["odometer_km"].describe())

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


In [11]:
# 3.1) Value counts - "price"-column

value_counts = autos["price"].value_counts()
value_counts_head = value_counts.sort_index(ascending= False).head(10)
value_counts_tail = value_counts.sort_index(ascending= True).head(10)
                                            
print(value_counts_head)
print("\n")
print(value_counts_tail)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: price, dtype: int64


0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
Name: price, dtype: int64


In [12]:
# 3.1) Value counts - "price"-column

value_counts = autos["odometer_km"].value_counts()
print(value_counts)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64


Oberservations:

|               | prize                   | odometer_km |
|---------------|-------------------------|-------------|
| Unique Values | 2.357                   | 13          |
| Min           | 0                       | 10.000      |
| Max           | 99.999.999              | 150.000     |
| Mean          | 9.840                   | 125.732     |
| Outliers      | 10.000.000 - 99.999.999 | None        |

In [13]:
# Remove outliers in "prize"-column

autos = autos[autos["price"] <= 10000000]

## Analyze time ranges in the five "date columns"

1. "date_crawled" - column

In [14]:
# - Extract the first ten characters, which represent the date
# - include missing values in the distribution and to use percentages instead of counts
# - rank by date in ascending order (earliest to latest)

date_date_crawled = autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending= True)
print(date_date_crawled)

2016-03-05    0.025384
2016-03-06    0.013942
2016-03-07    0.035965
2016-03-08    0.033265
2016-03-09    0.033205
2016-03-10    0.032124
2016-03-11    0.032485
2016-03-12    0.036765
2016-03-13    0.015562
2016-03-14    0.036625
2016-03-15    0.033985
2016-03-16    0.029504
2016-03-17    0.031524
2016-03-18    0.013062
2016-03-19    0.034905
2016-03-20    0.037825
2016-03-21    0.037505
2016-03-22    0.032925
2016-03-23    0.032385
2016-03-24    0.029104
2016-03-25    0.031744
2016-03-26    0.032485
2016-03-27    0.031044
2016-03-28    0.034845
2016-03-29    0.034185
2016-03-30    0.033625
2016-03-31    0.031904
2016-04-01    0.033805
2016-04-02    0.035405
2016-04-03    0.038685
2016-04-04    0.036525
2016-04-05    0.013102
2016-04-06    0.003180
2016-04-07    0.001420
Name: date_crawled, dtype: float64


Observations: Beside the latest two days and some few outliers the amount of crawled postings per day was pretty stable - around 3%

2. "ad_created" - column

In [15]:
date_ad_created = autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending= True)
print(date_ad_created)

2015-06-11    0.000020
2015-08-10    0.000020
2015-09-09    0.000020
2015-11-10    0.000020
2015-12-05    0.000020
                ...   
2016-04-03    0.038925
2016-04-04    0.036885
2016-04-05    0.011842
2016-04-06    0.003260
2016-04-07    0.001280
Name: ad_created, Length: 76, dtype: float64


Observations: The amount of ads created increased severly from the 05.03.2016 and remained pretty stable from there till the end of the dataset at the 01.04.2016. A reason for that could be the beginning of spring and hence the higher interest in selling and buying cars.

3. "last_seen" - column

In [16]:
date_last_seen = autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending= True)
print(date_last_seen)

2016-03-05    0.001080
2016-03-06    0.004421
2016-03-07    0.005361
2016-03-08    0.007581
2016-03-09    0.009841
2016-03-10    0.010762
2016-03-11    0.012522
2016-03-12    0.023803
2016-03-13    0.008981
2016-03-14    0.012802
2016-03-15    0.015882
2016-03-16    0.016442
2016-03-17    0.027924
2016-03-18    0.007421
2016-03-19    0.015742
2016-03-20    0.020703
2016-03-21    0.020723
2016-03-22    0.021583
2016-03-23    0.018583
2016-03-24    0.019563
2016-03-25    0.019203
2016-03-26    0.016962
2016-03-27    0.016022
2016-03-28    0.020863
2016-03-29    0.022343
2016-03-30    0.024843
2016-03-31    0.023823
2016-04-01    0.023103
2016-04-02    0.024903
2016-04-03    0.025364
2016-04-04    0.024623
2016-04-05    0.124297
2016-04-06    0.220991
2016-04-07    0.130938
Name: last_seen, dtype: float64


Observations: In correspondence with the increase of created adds in the beginning of march (see #2), can we also observe an increase in the "last_seen" category about seven days later (12.03.2016). Obviously, if there is an higher amount of postings, the amount of sales and hence deactivated offers should also increase. This cumulates in a very drastic increase in the beginning of April.

In [17]:
year_head_20 = autos["registration_year"].value_counts(dropna=False).sort_index(ascending= False).head(20)
year_tail_20 = autos["registration_year"].value_counts(dropna=False).sort_index(ascending= False).tail(20)
print(year_head_20)
print("\n")
print(year_tail_20)

9999       4
9996       1
9000       2
8888       1
6200       1
5911       1
5000       4
4800       1
4500       1
4100       1
2800       1
2019       3
2018     491
2017    1452
2016    1316
2015     399
2014     665
2013     806
2012    1323
2011    1634
Name: registration_year, dtype: int64


1953    1
1952    1
1951    2
1950    3
1948    1
1943    1
1941    2
1939    1
1938    1
1937    4
1934    2
1931    1
1929    1
1927    1
1910    9
1800    2
1500    1
1111    1
1001    1
1000    1
Name: registration_year, dtype: int64


Observations: The column contains some irregular values at the top (> 2016) and the bottom (< 1910)

## Filter the "registration_year" column


In [18]:
registration_year_unr = autos.loc[(autos["registration_year"] > 2016) | (autos["registration_year"] < 1910), "registration_year"]
registration_year_unr.shape

(1970,)

In [19]:
registration_year_unr.value_counts(normalize=True)

2017    0.737056
2018    0.249239
9999    0.002030
5000    0.002030
2019    0.001523
9000    0.001015
1800    0.001015
6200    0.000508
4500    0.000508
8888    0.000508
4800    0.000508
2800    0.000508
1001    0.000508
1000    0.000508
1111    0.000508
1500    0.000508
9996    0.000508
5911    0.000508
4100    0.000508
Name: registration_year, dtype: float64

Observation: 1970 entries (around 4% of the whole dataset) have unrealistic values in the column "registration_year". In my oppinion this amount of entries would still allow to erase them.

## Aggregation to understand the brand column

In [20]:
# Select the Top10 brands

brands_ori = autos["brand"].value_counts().head(10)
brands = brands_ori.index
print(brands)

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat'],
      dtype='object')


Calculate the mean price for the Top10 brands

In [21]:
# Create an empty dictionary to store our aggregate data

brand_mean_price = {}

# Loop over the unique values, and for each:
#     Subset the dataframe by the unique values
#     Calculate the mean of whichever column we're interested in
#     Assign the val/mean to the dict as k/v.

for b in brands:
    selected_rows = autos[autos["brand"] == b]
    mean = selected_rows["price"].mean()
    mean_round = "%.2f" % round(mean,2)
    brand_mean_price[b] = mean_round

print(brand_mean_price)

{'volkswagen': '5344.98', 'opel': '2845.91', 'bmw': '8252.92', 'mercedes_benz': '8389.94', 'audi': '8965.56', 'ford': '3913.02', 'renault': '2351.30', 'peugeot': '3010.87', 'fiat': '2697.68', 'seat': '4219.43'}


Observation: Cars from Mercedes Benz and Audio have the highest and from Renault and Fiat the lowest mean prices among the Top10 brands.

## Link between average mileage and mean price?

In [22]:
# Use the loop method from the last screen to calculate the mean mileage

brand_mean_mileage = {}

for b in brands:
    selected_rows = autos[autos["brand"] == b]
    mean = selected_rows["odometer_km"].mean()
    mean_round = "%.0f" % round(mean,0)
    brand_mean_mileage[b] = mean_round

print(brand_mean_mileage)

{'volkswagen': '128953', 'opel': '129295', 'bmw': '132522', 'mercedes_benz': '130882', 'audi': '129644', 'ford': '124153', 'renault': '128224', 'peugeot': '127352', 'fiat': '117012', 'seat': '122062'}


In [23]:
# Convert both dictionaries to series objects, using the series constructor.

bmp_series = pd.Series(brand_mean_price)
odo_series = pd.Series(brand_mean_mileage)
print(bmp_series)
print(odo_series)

volkswagen       5344.98
opel             2845.91
bmw              8252.92
mercedes_benz    8389.94
audi             8965.56
ford             3913.02
renault          2351.30
peugeot          3010.87
fiat             2697.68
seat             4219.43
dtype: object
volkswagen       128953
opel             129295
bmw              132522
mercedes_benz    130882
audi             129644
ford             124153
renault          128224
peugeot          127352
fiat             117012
seat             122062
dtype: object


In [24]:
# Create a dataframe from the first series object using the dataframe constructor.

mean_price_odometer = pd.DataFrame(bmp_series, columns=['mean_price'])
mean_price_odometer

mean_price
volkswagen       5344.98
opel             2845.91
bmw              8252.92
mercedes_benz    8389.94
audi             8965.56
ford             3913.02
renault          2351.30
peugeot          3010.87
fiat             2697.68
seat             4219.43

In [25]:
#Assign the other series as a new column in this dataframe.

mean_price_odometer.assign(average_mileage = odo_series)

mean_price average_mileage
volkswagen       5344.98          128953
opel             2845.91          129295
bmw              8252.92          132522
mercedes_benz    8389.94          130882
audi             8965.56          129644
ford             3913.02          124153
renault          2351.30          128224
peugeot          3010.87          127352
fiat             2697.68          117012
seat             4219.43          122062

Observation: There is no link between the mean price and the average mileage. Rather on the contrary - cars from Fiat have the lowest average mileage but also the lowest mean price, while cars from BMW, Audi and Merced-Benz (brands with highest mean prices) have on average run about 130.000 km. 